client/src/document/ingredients/browser-compatibility-table/feature-row.tsx

```diff
@@ -0,0 +1,382 @@
+import React, { useContext } from "react";
+import { Link } from "react-router-dom";
+import type bcd from "mdn-browser-compat-data/types";
+import { useAccordion, useAccordionParent } from "./accordion";
+import { BrowserInfoContext, BrowserName } from "./browser-info";
+import { asList, getFirst, isTruthy } from "./utils";
+
+function getSupportClassName(
+  support: bcd.SupportStatement | undefined
+): string {
+  if (!support) {
+    return "unknown";
+  }
+
+  let { version_added, version_removed, partial_implementation } = getFirst(
+    support
+  );
+
+  let className;
+  if (version_added === null) {
+    className = "unknown";
+  } else if (version_added) {
+    className = "yes";
+    if (version_removed) {
+      className = "no";
+    }
+  } else {
+    className = "no";
+  }
+  if (partial_implementation && !version_removed) {
+    className = "partial";
+  }
+
+  return className;
+}
+
+function StatusIcons({ status }: { status: bcd.StatusBlock }) {
+  const icons = [
+    status.experimental && {
+      title: "Experimental. Expect behavior to change in the future.",
+      text: "Experimental",
+      iconClassName: "ic-experimental",
+    },
+    status.deprecated && {
+      title: "Deprecated. Not for use in new websites.",
+      text: "Deprecated",
+      iconClassName: "ic-deprecated",
+    },
+    !status.standard_track && {
+      title: "Non-standard. Expect poor cross-browser support.",
+      text: "Non-standard",
+      iconClassName: "ic-non-standard",
+    },
+  ].filter(isTruthy);
+  return icons.length === 0 ? null : (
+    <div className="bc-icons">
+      {icons.map((icon) => (
+        <abbr key={icon.iconClassName} className="only-icon" title={icon.title}>
+          <span>{icon.text}</span>
+          <i className={icon.iconClassName} />
+        </abbr>
+      ))}
+    </div>
+  );
+}
+
+function NonBreakingSpace() {
+  return <>{"\u00A0"}</>;
+}
+
+function CellText({ support }: { support: bcd.SupportStatement | undefined }) {
+  const currentSupport = getFirst(support);
+
+  const added = currentSupport && currentSupport.version_added;
+  const removed = currentSupport && currentSupport.version_removed;
+
+  let status:
+    | { isSupported: "unknown" }
+    | { isSupported: "no" | "yes" | "partial"; label?: React.ReactNode };
```

Comment:
I have so much to learn! (not sure whether to put a smiley or a sad face emoji after that statement)

Reply:
yeah such acrobatics definitely take TS outside of the usual JS realm. If JS had enums like Rust or ReasonML this would be even more readable. In Rust one could express this as:
```rust
type Label = Option<React.Node>;
enum SupportStatus {
  Unknown,
  No(Label),
  Yes(Label),
  Partial(Label)
}
```
Which reads clearer to me. It basically says: status can have one of these values and for the non-`unknown` statuses it might have a label.